// uom/lib-cjs/unit-prefix.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Milli = exports.Centi = exports.Deci = exports.Hecto = exports.Kilo = exports.Mega = exports.Giga = void 0;
/* eslint-disable no-restricted-properties */
var Unit = require("./unit");
/////////////////
/// SI PREFIXES //
/////////////////
function Giga(name, u) {
    return Unit.timesNumber(name, Math.pow(10, 9), u);
}
exports.Giga = Giga;
function Mega(name, u) {
    return Unit.timesNumber(name, Math.pow(10, 6), u);
}
exports.Mega = Mega;
/**
 * Returns the specified unit multiplied by the factor <code>10<sup>3</sup></code>
 * @param unit Any unit.
 * @returns <code>unit.multiply(1e3)</code> .
 */
function Kilo(name, unit) {
    return Unit.timesNumber(name, Math.pow(10.0, 3), unit);
}
exports.Kilo = Kilo;
/**
 * Returns the specified unit multiplied by the factor <code>10<sup>2</sup></code>.
 * @param unit any unit.
 * @returns <code>unit.multiply(1e2)</code> .
 */
function Hecto(name, unit) {
    return Unit.timesNumber(name, Math.pow(10, 2), unit);
}
exports.Hecto = Hecto;
/**
 * Returns the specified unit multiplied by the factor <code>10<sup>-1</sup></code>.
 * @param unit any unit.
 * @return <code>unit.multiply(1e-1)</code>.
 */
function Deci(name, unit) {
    return Unit.timesNumber(name, Math.pow(10, -1), unit);
}
exports.Deci = Deci;
/**
 * Returns the specified unit multiplied by the factor <code>10<sup>-2</sup></code>.
 * @param unit any unit.
 * @returns <code>unit.multiply(1e-2)</code> .
 */
function Centi(name, unit) {
    return Unit.timesNumber(name, Math.pow(10, -2), unit);
}
exports.Centi = Centi;
/**
 * Returns the specified unit multiplied by the factor <code>10<sup>-3</sup></code>.
 * @param unit any unit. @return <code>unit.multiply(1e-3)</code> .
 */
function Milli(name, unit) {
    return Unit.timesNumber(name, Math.pow(10, -3), unit);
}
exports.Milli = Milli;
// export function Si<T extends Quantity>(toAdd: Unit.Unit<T>): Unit.Unit<T> {
//   // TODO
//   return toAdd;
// }
